Code/SimDivPickers/LeaderPicker.h

```diff
@@ -0,0 +1,427 @@
+                       double threshold, int nthreads) const {
+    CHECK_INVARIANT(distMat, "Invalid Distance Matrix");
+    if (!poolSize) throw ValueErrorException("empty pool to pick from");
+    if (poolSize < pickSize)
+      throw ValueErrorException("pickSize cannot be larger than the poolSize");
+    distmatFunctor functor(distMat);
+    return this->lazyPick(functor, poolSize, pickSize, firstPicks, threshold,
+                          nthreads);
+  }
+
+  /*! \overload */
+  RDKit::INT_VECT pick(const double *distMat, unsigned int poolSize,
+                       unsigned int pickSize) const {
+    RDKit::INT_VECT iv;
+    return pick(distMat, poolSize, pickSize, iv, default_threshold,
+                default_nthreads);
+  }
+};
+
+#ifdef USE_THREADED_LEADERPICKER
+template <typename T>
+void *LeaderPickerWork(void *arg);
+
+template <typename T>
+struct LeaderPickerState {
+  typedef struct {
+    int *ptr;
+    unsigned int capacity;
+    unsigned int len;
+    unsigned int next[2];
+  } LeaderPickerBlock;
+  typedef struct {
+    LeaderPickerState<T> *stat;
+    pthread_t tid;
+    unsigned int id;
+  } LeaderPickerThread;
+
+  std::vector<LeaderPickerThread> threads;
+  std::vector<LeaderPickerBlock> blocks;
+  pthread_barrier_t wait;
+  pthread_barrier_t done;
+  std::vector<int> v;
+  LeaderPickerBlock *head_block;
+  unsigned int thread_op;
+  unsigned int nthreads;
+  unsigned int tick;
+  double threshold;
+  int query;
+  T *func;
+
+  LeaderPickerState(unsigned int count, int nt) {
+    v.resize(count);
+    for (unsigned int i = 0; i < count; i++) v[i] = i;
+
+    // InitializeBlocks
+    unsigned int bcount;
+    unsigned int bsize;
+    if (nt > 1) {
+      bsize = 4096;
+      bcount = (count + (bsize - 1)) / bsize;
+      unsigned int tasks = (bcount + 1) / 2;
+      // limit number of threads to available work
+      if (nt > (int)tasks) nt = tasks;
+    } else {
+      bsize = 32768;
+      bcount = (count + (bsize - 1)) / bsize;
+    }
+    blocks.resize(bcount);
+    head_block = &blocks[0];
+    tick = 0;
+
+    if (bcount > 1) {
+      int *ptr = &v[0];
+      unsigned int len = count;
+      for (unsigned int i = 0; i < bcount; i++) {
+        LeaderPickerBlock *block = &blocks[i];
+        block->ptr = ptr;
+        if (len > bsize) {
+          block->capacity = bsize;
+          block->len = bsize;
+          block->next[0] = i + 1;
+        } else {
+          block->capacity = len;
+          block->len = len;
+          block->next[0] = 0;
+          break;
+        }
+        ptr += bsize;
+        len -= bsize;
+      }
+    } else {
+      head_block->capacity = count;
+      head_block->len = count;
+      head_block->next[0] = 0;
+      head_block->next[1] = 0;
+      head_block->ptr = &v[0];
+    }
+
+    // InitializeThreads
+    if (nt > 1) {
+      nthreads = nt;
+      pthread_barrier_init(&wait, NULL, nthreads + 1);
```

Comment:
Why is this nthreads + 1?

Reply:
I haven't spent much time looking at this since we've got it disabled. 
I think leaving it disabled until post-release is the way to go

Reply:
Conceptually there are "nthreads" worker threads plus the master thread that launches them.
On starting up, all the worker threads immediately block on the "wait" barrier, effectively waiting for the master to join (hence "nthreads+1") to release them all.  The master thread then blocks on the "done" barrier, waiting for all the "nthreads" workers to complete before continuing.  Hence this pair of barriers provides an efficient implementation of the "bulk synchronous parallel (BSP)" or symmetric co-routines parallel design pattern.  Greg has a copy of a small pthreads benchmark that confirms that this is more efficient than (the more commonly encounted) repeated pthread_create/pthread_join loops (for example, this idiom is used in GNU's OpenMP implementation).

I completely agree there should be a comment in the code here, explaining this.

Reply:
Thanks roger, I missed the master call on wait.